leaderelection: Allow leader elected code to step down on a context cancel

smarterclayton opened the pull request:
The current code simply exits without continuing to renew the lease, which means participants using a slower lease duration might have to wait multiple minutes before a new leader is elected. Allow an optional flag to be set on LeaderElectionConfig that will release the lease when the calling context is cancelled. Callers *must* ensure their lease guarded code has exited before the context is cancelled, or other processes may acquire the lease before this lease has released (this is why the behavior is opt-in).

Add an example command that demonstrates how cancellation could be done.

This supports operators, custom controllers, and eventually our own controllers in being able to rapidly react when they are updated (i.e. you update your operator via a helm update, but you don't want to have to wait until the lease expires).

/kind feature


```release-note
The leaderelection package allows the lease holder to release its lease when the calling context is cancelled. This allows
faster handoff when a leader-elected process is gracefully terminated. 
```

smarterclayton committed:
leaderelection: Allow leader elected code to step down on a context cancel

The current code simply exits without continuing to renew the lease, which means
participants using a slower lease duration might have to wait multiple minutes
before a new leader is elected. Allow an optional flag to be set on
LeaderElectionConfig that will release the lease when the calling context is
cancelled. Callers *must* ensure their lease guarded code has completed before
the context is cancelled, or other processes may acquire the lease before this
lease has released.

Add an example command that demonstrates how cancellation could be done.

As a convenience to users, make event recorder optional - not all users of the
lock code will need a recorder.

smarterclayton commented:
/assign @mikedanese

smarterclayton commented:
@deads2k as well

justinsb commented on staging/src/k8s.io/client-go/tools/leaderelection/example/main.go:
```diff
@@ -0,0 +1,106 @@
+		// you call cancel. Otherwise, you could have a background
+		// loop still running and another process could
+		// get elected before your background loop finished, violating
+		// the stated goal of the lease.
```
This seems like a fairly hard guarantee to make.  Do we have an example, or are you planning a follow-on PR to make this easier?

smarterclayton replied:
There's a couple of options

1. most controllers are reentrant, so it doesn't actually matter (the ones I would be concerned with in controller manager are the ones that talk to external systems and allocate things, and even they are probably ok)
2. have a global "disable client" option when you get a client (basically a .Stop() on client) which is pretty easy to make work - this would be the best option for most people
3. some controllers are correctly written even today to drain, it's reasonably easy to test them

The ones who would consume this first are things like 1.  I need to look at 2.

smarterclayton replied:
I implemented 2 - it should be fairly easy to use (I made the example demonstrate it)

smarterclayton commented on staging/src/k8s.io/client-go/tools/leaderelection/example/main.go:
```diff
@@ -0,0 +1,106 @@
+		LeaseDuration:   60 * time.Second,
+		RenewDeadline:   15 * time.Second,
+		RetryPeriod:     5 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
```
Only if exit of the run loop guarantees drain (some do, some don't)

smarterclayton commented on staging/src/k8s.io/client-go/tools/leaderelection/example/main.go:
```diff
@@ -0,0 +1,106 @@
+		config, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		log.Fatalf("failed to create client: %v", err)
```
I can flip the example, not hard.  we don't do a ton of examples, but I wouldn't expect someone looking at this to have to use klog.

justinsb replied:
Right - this was more an action-item for myself as someone wanting to get involved in klog :-)

smarterclayton commented:
/retest

mikedanese commented on staging/src/k8s.io/client-go/tools/leaderelection/example/main.go:
```diff
@@ -0,0 +1,122 @@
+		},
+		Client: kubernetes.NewForConfigOrDie(config).CoreV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: id,
```
Doesn't this need to be unique per instance of this main?

smarterclayton replied:
Yes, as a test example you have to pass that in as an argument.

mikedanese commented on staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go:
```diff
@@ -249,6 +256,28 @@ func (le *LeaderElector) renew(ctx context.Context) {
+	leaderElectionRecord := rl.LeaderElectionRecord{
+		LeaderTransitions: le.observedRecord.LeaderTransitions,
+	}
+	if err := le.config.Lock.Update(leaderElectionRecord); err != nil {
```
tryAcquireOrRenew calls Get() which refreshes resource version. We might want to refresh here as well to avoid conflict errors.

smarterclayton replied:
So when writing this I was biasing to "best effort" and "step down as quickly as possible before we get nuked by the sig kill".  If we refresh, we could potentially have to wait 2 round trips vs one, with the potential for longer behavior.

Since we only step down if we're the leader, shouldn't we have an updated resource version in the lock already as the last writer?

mikedanese replied:
> shouldn't we have an updated resource version in the lock already as the last writer?

Usually. Tradeoff SGTM.

mikedanese commented on staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go:
```diff
@@ -284,7 +313,8 @@ func (le *LeaderElector) tryAcquireOrRenew() bool {
 		le.observedTime = le.clock.Now()
 	}
-	if le.observedTime.Add(le.config.LeaseDuration).After(now.Time) &&
+	if len(oldLeaderElectionRecord.HolderIdentity) > 0 &&
```
What are the compatibility implications here?

smarterclayton replied:
I'll add a test to prove behavior for this, but older clients watching observe the lease as still held (they don't refresh).  So they assume the holder still holds.  If a new client comes up (N+1) while old clients (N) are still running, the new client would have to acquire the lease, then step down, and the old clients would simply be less aggressive in acquiring.

smarterclayton committed:
Make wrapping a client transport more pleasant

Properly wrapping a transport can be tricky. Make the normal case
(adding a non-nil transport wrapper to a config) easier with a helper.
Also enforce a rough ordering, which in the future we can use to
simplify the WrapTransport mechanism down into an array of functions
we execute in order and avoid wrapping altogether.

smarterclayton committed:
Add transport wrapper that blocks api calls after context close

The ContextCanceller transport wrapper blocks all API requests
after the provided context is closed. Used with the leader election
step down, a controller can ensure that new requests are not made
after the client has stepped down.

smarterclayton commented:
Additional changes (rounding out the "how do I use this practically from a controller").

* refactor how transport wrappers are provided so that we can simplify the client - it's fairly error prone
* fixed a minor bug where we weren't copying the rest.Config passed to the discovery client, so we could unintentionally mutate the rest.Config for a subsequent call
* added a `ContextCircuitBreaker` transport wrapper that prevents new API requests from being made once the context is closed.  This ensures that no new api requests are made once we release the lease, although it's possible that requests based on an earlier observed state are still waiting to be sent (which should be safe).

smarterclayton commented:
Ready for review as a complete "a controller can easily step down" PR now.

smarterclayton commented:
/retest

mikedanese commented on staging/src/k8s.io/client-go/tools/leaderelection/leaderelection_test.go:
```diff
@@ -208,83 +241,86 @@ func testTryAcquireOrRenew(t *testing.T, objectType string) {
-		var reportedLeader string
-		var lock rl.Interface
+	for i := range tests {
+		test := &tests[i]
```
Is this necessary? I tooks like a _, t reference won't change out from underneath the test.

smarterclayton replied:
if it ever gets made parallel it would be - I can change if you like.

mikedanese commented on staging/src/k8s.io/client-go/transport/transport.go:
```diff
@@ -167,3 +168,60 @@ func rootCertPool(caData []byte) *x509.CertPool {
+// ContextCircuitBreaker prevents new requests after the provided context is finished.
+// err is returned when the context is closed, allowing the caller to provide a context
+// appropriate error.
+func ContextCircuitBreaker(ctx context.Context, err error) WrapperFunc {
```
ContextCanceller or Cancellable? When I here circuit breaker, I think of error thresholds.

smarterclayton replied:
Fair.  Stopper or Disable also work.  I like Cancellable for succinctness.

smarterclayton replied:
Went with ContextCanceller since it read well

mikedanese commented:
mostly lgtm.

mikedanese commented:
/lgtm
